azblob/zc_policy_retry.go

```diff
@@ -296,6 +296,12 @@ func (rc *contextCancelReadCloser) Close() error {
 	return err
 }
 
+func (rc *contextCancelReadCloser) CancelRequest() {
```

Comment:
Instead of adding a new receiver, we could potentially just call `Close()`. It cancels the context in addition to closing the connection, which is desired.

Reply:
If we closed the connection while another go-routine is reading from the body, I'd expect the `Read` function to return an error and say that the connection broke, but certain bytes were read. This should be double-checked.

Reply:
I tried that, expecting the same behavior that you mentioned. And the first time it worked great... But, the next time I ran it I got about 6 "context canceled" errors.  I suspected some kind of race condition, but did not identify exactly where it might have been.  That might be debuggable and fixable but..

.. then I realized something else.  If we use `Close()`, we're actually making an assumption there. We are assuming that (a) closing this particular response Body will actually cancel the context and (b) the context that gets cancelled is one that was created uniquely for this request (i.e. it's a child context, not a parent one).  Those assumptions are true for bodies that are implemented by `contextCancelReadCloser` but not necessarily true for all hypothetical response bodies.  So... using the explicit RequestCanceller interface provides a check that our assumptions are met.  Why is that check useful?  Because someone may one day use retryReader without policyRetry, and the assumptions would then be no longer met.  With the explicit reliance on `contextCancelReadCloser` such a code change would be immediately detected as an error - because suddenly retryReader would be using a request body that did not implement the interface.  We can't get the same effect with `Close()` because _all_ bodies are closeable.

Reply:
@zezha-msft I've made and pushed the changes to simplify it.  As discussed offline, and documented in the new commit, early Closing does appear to be safe. It works independently of the context cancel mechanism, but looking through the Go SDK source code indicates that it has the same effect.